### PR TITLE
Analytics

### PR DESCRIPTION
**Pull request by sgkang**
Electric Dipole Frequency Doamin

### CODE REVIEW

File: SimPEG/EM/Analytics/FDEM_fields.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
+from __future__ import division
+import numpy as np
+from scipy.constants import mu_0, pi
+from scipy.special import erf
+from SimPEG import Utils
+
+def E_from_ElectricDipoleWholeSpace(XYZ, srcLoc, sig, f, current=1., length=1., orientation='X', mu=mu_0):
```
Comment (lheagy) [inline SimPEG/EM/Analytics/FDEM_fields.py]: docs?? :)

File: SimPEG/EM/Analytics/FDEM_fields.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    elif orientation.upper() == 'Z':
+        # x --> z, y --> x, z --> y
+        Ez = front*((dz**2 / r**2)*mid + (k**2 * r**2 -1j*k*r-1.))
+        Ex = front*(dz*dx  / r**2)*mid
+        Ey = front*(dz*dy  / r**2)*mid
+        return Ex, Ey, Ez
+
+
+def E_galvanic_from_ElectricDipoleWholeSpace(XYZ, srcLoc, sig, f, current=1., length=1., orientation='X', mu=mu_0):
+    epsilon = 8.854187817*(10.**-12)
```
Comment (lheagy) [inline SimPEG/EM/Analytics/FDEM_fields.py]: `from scipy.sparse import epsilon_0`

File: SimPEG/EM/Analytics/FDEM_fields.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    omega = 2.*np.pi*f
+    sig_hat = sig + 1j*omega*epsilon
+
+    XYZ = Utils.asArray_N_x_Dim(XYZ, 3)
+    # Check
+    if XYZ.shape[0] > 1 & f.shape[0] > 1:
+        raise Exception("I/O type error: For multiple field locations only a single frequency can be specified.")
+
+    dx = XYZ[:,0]-srcLoc[0]
```
Comment (lheagy) [inline SimPEG/EM/Analytics/FDEM_fields.py]: looks like this and the following 2 lines are used in a number of places, perhaps turn it into a small function?

```
def getSrcCenteredGrid(XYZ, srcLoc):
    # center the grid
    return XYZcentered
```

File: SimPEG/EM/Analytics/FDEM_fields.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+def J_inductive_from_ElectricDipoleWholeSpace(XYZ, srcLoc, sig, f, current=1., length=1., orientation='X', mu=mu_0):
+    Ex_inductive, Ey_inductive, Ez_inductive = E_inductive_from_ElectricDipoleWholeSpaced(XYZ, srcLoc, sig, f, current=1., length=1., orientation='X', mu=mu_0)
+    Jx_inductive = sig*Ex_inductive
+    Jy_inductive = sig*Ey_inductive
+    Jz_inductive = sig*Ez_inductive
+    return Jx_inductive, Jy_inductive, Jz_inductive
+
+
+def H_from_ElectricDipoleWholeSpace(XYZ, srcLoc, sig, f, current=1., length=1., orientation='X', mu=mu_0):
+    epsilon = 8.854187817*(10.**-12)
```
Comment (lheagy) [inline SimPEG/EM/Analytics/FDEM_fields.py]: import from scipy

File: SimPEG/EM/Analytics/FDEM_fields.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        # x --> z, y --> x, z --> y
+        Ez_galvanic = front*((dz**2 / r**2)*mid + (-1j*k*r-1.))
+        Ex_galvanic = front*(dz*dx  / r**2)*mid
+        Ey_galvanic = front*(dz*dy  / r**2)*mid
+        return Ex_galvanic, Ey_galvanic, Ez_galvanic
+
+
+def E_inductive_from_ElectricDipoleWholeSpace(XYZ, srcLoc, sig, f, current=1., length=1., orientation='X', mu=mu_0):
+    epsilon = 8.854187817*(10.**-12)
+    omega = 2.*np.pi*f
```
Comment (lheagy) [inline SimPEG/EM/Analytics/FDEM_fields.py]: looks like this is used a few places - could define it as a lambda function (`omega = lambda f: 2.*np.pi*f`) outside of this . docs: https://docs.python.org/2/tutorial/controlflow.html#lambda-expressions

File: SimPEG/EM/Analytics/FDEM_fields.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    XYZ = Utils.asArray_N_x_Dim(XYZ, 3)
+    # Check
+    if XYZ.shape[0] > 1 & f.shape[0] > 1:
+        raise Exception("I/O type error: For multiple field locations only a single frequency can be specified.")
+
+    dx = XYZ[:,0]-srcLoc[0]
+    dy = XYZ[:,1]-srcLoc[1]
+    dz = XYZ[:,2]-srcLoc[2]
+
+    r  = np.sqrt( dx**2. + dy**2. + dz**2.)
```
Comment (lheagy) [inline SimPEG/EM/Analytics/FDEM_fields.py]: same lambda fct comments for `r`, `k`

File: SimPEG/EM/Analytics/__init__.py [hunk 7]
```diff
@@ -2,3 +2,4 @@
 from FDEM import hzAnalyticDipoleF
 from FDEMcasing import *
 from DC import DCAnalyticHalf, DCAnalyticSphere
+from FDEM_fields import *
```
Comment (lheagy) [inline SimPEG/EM/Analytics/__init__.py]: are all of the functions in this file specific to dipoles? Perhaps we could name the file `FDEMDipolarFields.py` or something along those lines

**Comment by lheagy** [comment]
Exciting to have more analytics in SimPEG! I have left a few comments - let me know if there is anything you would like to chat about or it anything needs clarification

File: SimPEG/EM/Analytics/FDEM_fields.py (hunk 2 above)
Comment (sgkang) [inline SimPEG/EM/Analytics/FDEM_fields.py]: made it as a variable: epsilon and mu, but input is kappa (susceptibility) and epsr (relative permittivity)

**Pull request merged**